File: Emulator/src/tests/ROM.ts
import test from 'ava'
import { ROM } from '../components/ROM'

test('ROM can init', t => {
	t.not(new ROM(), null)
})

test('ROM can read', t => {
	const rom = new ROM()

	t.is(rom.read(0x0000), 0xEA)
	t.is(rom.read(0x7FFF), 0xEA)
})

test('ROM can load', t => {
	const rom = new ROM()

	rom.load([...Array(ROM.SIZE)].fill(0xAA))

	t.is(rom.data[0x0000], 0xAA)
	t.is(rom.data[0x7FFF], 0xAA)

	rom.load([...Array(ROM.SIZE)].fill(0x55))

	t.is(rom.data[0x0000], 0x55)
	t.is(rom.data[0x7FFF], 0x55)

	// Test loading with oversize data fails to load
	rom.load([...Array(ROM.SIZE + 1)].fill(0xEA))

	t.not(rom.data[0x0000], 0xEA)
	t.not(rom.data[0x7FFF], 0xEA)
	t.is(rom.data[0x0000], 0x55)
	t.is(rom.data[0x7FFF], 0x55)
})

File: Firmware/DB Emulator/lib/ROM/ROM.cpp
#include "ROM.h"

ROM::ROM(uint16_t size) {
  this->size = size;
  this->data = new uint8_t[size];

  for (uint16_t a = 0x0000; a < size; a++) {
    this->data[a] = 0xEA; // Load with NOPs
  }
}

ROM::~ROM() {
  delete[] this->data;
}

uint8_t ROM::read(uint16_t index) {
  return this->data[index];
}

void ROM::write(uint16_t index, uint8_t value) { /* Do nothing */ }